File: Python/documentation.py
# ________________________________________________________________________
# LICENCIATURA EN CIENCIAS DE LA COMPUTACION - R123 Programacion II - 2022
# 
# Documentacion Trabajo Practico Final - Parte Python
# Alumno: Sferco Martin
# ________________________________________________________________________
#
#
# - EJECUCION DEL PROGRAMA:
#                           Para ejecutar el programa, debemos abrir una consola dentro
#   de la carpeta Python/, la cual se encuentra en la raiz del proyecto. Luego, ejecutamos
#   el siguiente comando:
#   
#           $ python3 main.py ../generados/<nombreArchivo> <colorJugador> <nivelDificultad>
#           
#   en donde todos los argumentos deben estar separados por un espacio. Cada uno 
#   de los argumentos nombreArchivo, turnoJugador y nivelDificultad los debera elegir el
#   usuario y pueden tomar los siguientes valores:
#
#       - <nombreArchivo>       Es el nombre del archivo del cual vamos a empezar la partida.
#                             Tiene que incluir la terminacion .txt y debe estar dentro de
#                             la carpeta generados/.
#   
#       - <colorJugador>        Es el color del jugador, y puede ser B o N. 
#
#       - <nivelDificultad>     Es el nivel de dificultad de la maquina y puede ser 0 o 1.
#
#   En el caso de que estemos en Windows, debemos utilizar python en lugar de python3.
#   En el caso de que se quiera utilizar otra carpeta con archivos generados, se tiene 
#   que incluir dicha carpeta en la raiz del proyecto y se debe ejecutar el comando de 
#   la siguiente manera:
#
#           $ python3 main.py ../<tuCarpeta>/<nombreArchivo> <colorJugador> <nivelDificultad>
#
#   reemplazando <tuCarpeta> por el nombre de la carpeta que hayas incluido en la raiz
#   del proyecto.
#
#
# - EJECUCION DE LOS TESTS:
#                       Para ejecutar los tests de las funciones del programa, debemos abrir
#   una terminal en la carpeta Python/, la cual se encuentra en la raiz del proyecto. A 
#   continuacion, ejecutamos el siguiente comando.
#
#           $ python3 -m pytest test.py
#
#   En el caso de que estemos en Windos, debemos utilizar python en lugar de python3.
#   
#
# - DIVISION DE ARCHIVOS:
#                        Dividimos la totalidad del programa en varios archivos
#   para facilitar el trabajo y la lectura del mismo en un futuro:
#
#       - main.py       Aqui agrupamos todas las funciones y es donde ejecutaremos
#                       el programa.
#
#       - test.py       Aqui escribimos todos los test de las funciones.
#
#       - functions.py  Aqui escribimos todas las funciones relacionadas con el
#                       funcionamiento del programa. Luego son agrupadas en el 
#                       main.py.
#       
#       - show.py       Se encuentran todas las funciones que se encargan de 
#                       mostrar mensajes importantes o informacion por pantalla.
#
#       - arguments.py  Se encuentran todas las funciones que se encargan de verificar
#                       que todos los argumentos pasados por el jugador sean correctos.
# 
#  
# - DISENIO DE DATOS Y RESOLUCION:
#                                  A continuacion, explicaremos un poco como se diseniaron los 
#   los distintos tipos de datos y como trabajamos con la resolucion.
#   
#   
#       - Casillas:
#                   el tablero utiliza letras y numeros para representar las coordenadas de una 
#       casilla. Esto lo normalizamos realizando una asociacion entre las coordenadas reales en
#       en el tablero, con una tupla de enteros. Las columnas de la A a la H las asociamos con
#       numeros del 0 al 7, y con las filas del 1 al 8, tambien las asociamos con numeros del 
#       1 al 7. Algunos ejemplos de estas asociaciones serian:
#
#           "A1" -> (0,0)
#           "H8" -> (7,7)
#           "B3" -> (1,2)
#
#       - Tablero:
#                 Al tablero lo representamos de manera implicita. Utilizaremos un diccionario
#       de conjuntos y solo guardaremos aquellas fichas que se encuentren ocupadas. Las que no
#       se encuentren ocupadas, no las incluiremos en dicho diccionario.
#
#       - Analisis de archivo leido:
#                                   Como estamos suponiendo que el archivo pasado por linea de 
#       comandos es un archivo correcto, no verificamos la informacion que posee dentro. Lo 
#       consideramos correcto ya que fue generado por el programa en C.

File: C/documentation.c
// ________________________________________________________________________
// LICENCIATURA EN CIENCIAS DE LA COMPUTACION - R123 Programacion II - 2022
//
// Documentacion Trabajo Practico Final - Parte C
// Alumno: Sferco Martin
// ________________________________________________________________________
//
//
// - EJECUCION DEL PROGRAMA:
//                          Para compilar el programa debemos entrar a la 
//   carpeta C/ que se encuentra en la raiz del proyecto. Luego abrimos una
//   terminal dentro de dicha carpeta y ejecutamos el siguiente comando:
//
//          $ gcc main.c functions.c -o reversi
// 
//  Una vez compilado, ejecutamos el programa con el siguiente comando:
// 
//          $ ./reversi ../juegosPrueba/nombreArchivo 
//  
//  en donde nombreArchivo debe colocarse el nombre de archivo de juego que 
//  quieras procesar (con la extension .txt), que se encuentra dentro de la
//  carpeta /juegosPrueba. En el caso que se quiera utilizar otra carpeta con
//  archivos de prueba deberiamos utilizar el siguiente comando:
//
//          $ ./reversi ../tuCarpeta/nombreArchivo 
//
//  reeemplazando tuCarpeta por el nombre de la carpeta que hayas agregado
//  en la raiz del proyecto.
//
//
//  - EJECUCION DE LOS TESTS:
//                           Para compilar los archivos de tests, primero
//    debemos entrar a la carpeta C/ que se encuentra en la raiz del del 
//    proyecto. Luego abrimos una terminal dentro de dicha carpeta y 
//    ejecutamos el siguiente comando:
//
//          $ gcc test.c functions.c -o test
//
//    Una vez compilado, ejecutamos los test con el siguiente comando:
//
//          $ ./test
//
//  
//  - DISENIO DE DATOS Y RESOLUCION:
//                                  A continuacion, explicaremos un poco el
//    disenio de datos que hemos utilizado y un poco sobre la resolucion:
//
//          - Jugadores:
//                      Para los jugadores creamos una estructura llamada
//            Jugador, la cual tiene un campo de tipo char* donde guardaremos
//            de manera dinamica los nombres, y otro campo de tipo char donde
//            guardaremos el color del jugador.
//
//          - Tablero:       
//                    La informacion de las fichas jugadas de cada color, la
//            iremos almacenanando en un dato de tipo char[][], para que podamos
//            tener una idea de coordenadas en X e Y. Cada una de las 
//            casillas tendra tres posibles valores:
//
//                  - 'X' - No hay ninguna ficha colocada.
//                  - 'B' - Hay una ficha blanca colocada.
//                  - 'N' - Hay una ficha negra colocada.
//
//          - Casillas:
//                      Para ir pasando la informacion de las casillas entre 
//          funciones, trabajamos con una estructura llamada Casilla, la cual
//          tiene dos campos de tipo entero, las cuales guardan la posicion
//          X e Y de cada una de las casillas.
//
//          - Informacion preliminar:
//                                   Antes de comenzar con la lectura del 
//          archivo, verificaremos que los colores de los dos jugadores
//          sean correctos y diferentes entre si. Ademas, tenemos que ver
//          que el color de inicio sea correcto. Una vez que se haya 
//          verificado toda esta informacion, la almacenamos en variables
//          y ya podemos comenzar la verifiacion de las jugadas.
//
//          - Terminacion de partida:
//                                  El programa puede finalizar por 
//          diferentes razones:
//
//              - Error en jugada:
//
//                  - No cumple el formato estipulado.
//                  - Se sale de rango.
//                  - La casilla se encuentra ocupada.
//                  - La jugada no genera cambios.
//                  - Se salta el turno cuando habia jugadas posibles.
//      
//              - Finalizacion de lectura:
//                                       Esto indica que se llego al final
//              del archivo y no se se proporcionaron mas jugadas.
//
//              - Juego terminado: 
//
//                  - Se colocaron todas las fichas.
//                  - Alguno de los jugadores se quedo sin fichas.
//                  - Se salto el turno dos veces consecutivas.
//
//
//          - Generacion de archivo: 
//                                  En el caso de que la partida haya
//          quedado a medias y se pueda seguir jugando, el programa
//          generara un archivo .txt en la carpeta juegosGenerados/, la
//          cual se encuentra en la raiz del proyecto. Dicho archivo
//          contendra informacion de la partida, la cual sera la posicion
//          de todas las fichas y el turno del jugador que le toca en el
//          siguiente turno.
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
